File: solutions/go/dnd-character/1/dnd_character.go
package dndcharacter

import (
    "math/rand"
    "math"
)

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
    modifier := (float64(score) - 10.0) / 2.0
    roundDown := int(math.Floor(modifier))
    return roundDown
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
    totalDices := 4
    diceFaces := 6
	dices := make([]int, totalDices)
    for i := 0; i < totalDices; i++ {
        dices[i] = rand.Intn(diceFaces) + 1 
    }
    lower := dices[0]
    total := 0
    for _, v := range dices {
        if v < lower {
            lower = v
        }
        total += v
    }
    return total - lower
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	character := Character {
        Strength: Ability(),
        Dexterity: Ability(),
        Constitution: Ability(),
        Intelligence: Ability(),
        Wisdom: Ability(),
        Charisma: Ability(),
        Hitpoints: 10,
    }

	character.Hitpoints += Modifier(character.Constitution)
    
    return character
}

File: solutions/csharp/dnd-character/1/DndCharacter.cs
public class DndCharacter(int strength, int dexterity, int constitution, int intelligence, int wisdom, int charisma, int hitpoints)
{
    public int Strength { get; } = strength;
    public int Dexterity { get; } = dexterity;
    public int Constitution { get; } = constitution;
    public int Intelligence { get; } = intelligence;
    public int Wisdom { get; } = wisdom;
    public int Charisma { get; } = charisma;
    public int Hitpoints { get; } = hitpoints;

    public static int Modifier(int score) => (int)Math.Floor((score - 10) / 2.0);

    public static int Ability()
    {
        var rolls = new List<int>();
        var random = new Random();
        for (int i = 0; i < 4; i++)
        {
            rolls.Add(random.Next(1, 7));
        }
        rolls.Remove(rolls.Min());
        return rolls.Sum();
    }

    public static DndCharacter Generate()
    {
        var constitution = Ability();
        return new DndCharacter(Ability(), Ability(), constitution, Ability(), Ability(), Ability(), 10 + Modifier(constitution));
    } 
}
